File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

class Solution {
    
    static class Node {
        final int x, y, v;
        Node left, right;
        Node(int x, int y, int v) {
            this.x = x;
            this.y = y;
            this.v = v;
        }
    }

    private static void insert(Node root, Node node) {
        Node cur = root;
        while (true) {
            if (node.x < cur.x) {
                if (cur.left == null) {
                    cur.left = node;
                    return;
                }
                cur = cur.left;
            } else {
                if (cur.right == null) {
                    cur.right = node;
                    return;
                }
                cur = cur.right;
            }
        }
    }

    private static void preOrder(Node root, List<Integer> result) {
        if (root == null) return;
        Deque<Node> stack = new ArrayDeque<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            result.add(node.v);
            if (node.right != null) stack.push(node.right);
            if (node.left != null) stack.push(node.left);
        }
    }

    private static void postOrder(Node root, List<Integer> result) {
        if (root == null) return;
        Deque<Node> stack = new ArrayDeque<>();
        LinkedList<Integer> output = new LinkedList<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            output.addFirst(node.v);
            if (node.left != null) stack.push(node.left);
            if (node.right != null) stack.push(node.right);
        }
        result.addAll(output);
    }

    public int[][] solution(int[][] nodeinfo) {
        Node[] nodes = new Node[nodeinfo.length];
        for (int v = 0; v < nodeinfo.length; v++) {
            nodes[v] = new Node(nodeinfo[v][0], nodeinfo[v][1], v + 1);
        }

        Arrays.sort(nodes, (n1, n2) ->
                n1.y != n2.y ? n2.y - n1.y : n1.x - n2.x);

        Node root = nodes[0];
        for (int i = 1; i < nodes.length; i++) {
            insert(root, nodes[i]);
        }

        List<Integer> preResult = new ArrayList<>();
        List<Integer> postResult = new ArrayList<>();
        preOrder(root, preResult);
        postOrder(root, postResult);

        return new int[][] {
                preResult.stream().mapToInt(v -> v).toArray(),
                postResult.stream().mapToInt(v -> v).toArray()
        };
    }
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys
sys.setrecursionlimit(10_000)

class Node:
    def __init__(self, x, y, k):
        self.x = x
        self.y = y
        self.k = k
        self.left = None
        self.right = None
        
        
class Tree:
    def __init__(self):
        self.root = None
    
    def insert(self, x, y, k):
        self.root = self._insert(self.root, x, y, k)
        return self.root is not None
    
    def _insert(self, node, x, y, k):
        if node is None:
            return Node(x, y, k)
        if x < node.x:
            node.left = self._insert(node.left, x, y, k)
        else:
            node.right = self._insert(node.right, x, y, k)
        return node
    
    def preorder(self):
        return self._preorder(self.root, [])
    
    def _preorder(self, node, result):
        if node:
            result.append(node.k)
            self._preorder(node.left, result)
            self._preorder(node.right, result)
        return result
    
    def postorder(self):
        return self._postorder(self.root, [])
    
    def _postorder(self, node, result):
        if node:
            self._postorder(node.left, result)
            self._postorder(node.right, result)
            result.append(node.k)
        return result


def solution(nodeinfo):
    tree = Tree()
    
    nodes = [(x, y, i) for i, (x, y) in enumerate(nodeinfo, 1)]
    nodes.sort(key=lambda n: (-n[1], n[0]))
    
    for x, y, i in nodes:
        tree.insert(x, y, i)
    
    return tree.preorder(), tree.postorder()
